### PR TITLE
feat: campaign setup and first four properties

### PR DESCRIPTION
**Pull request by 0xDiscotech**
CLOSES OPT-619

### CODE REVIEW

**Comment by 0xDiscotech** [comment]
@hexshire could you create the Mesuda script on the `package.json` or `justfile` with the needed configs (if any)

File: packages/contracts-bedrock/test/invariants/interop/utils/Deployer825.sol [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    function deploySuperchainERC20() public returns (address superchainERC20) {
+        superchainERC20 = address(new MedusaToken());
+    }
+
+    function deploySuperchainTokenBridge() public returns (address superchainTokenBridge) {
+        superchainTokenBridge = address(new SuperchainTokenBridge());
+    }
+}
+
+contract MedusaToken is SuperchainERC20 {
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/invariants/interop/utils/Deployer825.sol]: Let's call it SuperToken
```suggestion
contract SuperToken is SuperchainERC20 {
```

File: packages/contracts-bedrock/test/invariants/interop/Setup.sol [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    ICrossL2Inbox constant inbox = ICrossL2Inbox(Predeploys.CROSS_L2_INBOX);
+    IL2ToL2CrossDomainMessenger constant messenger =
+        IL2ToL2CrossDomainMessenger(Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER);
+    ISuperchainERC20 immutable medusaToken;
+    ISuperchainTokenBridge constant tokenBridge = ISuperchainTokenBridge(Predeploys.SUPERCHAIN_TOKEN_BRIDGE);
+
+    // Actors
+    address immutable guardian = vm.addr(uint256(keccak256("Guardian")));
+    address immutable upgrader = vm.addr(uint256(keccak256("Upgrader")));
+    address immutable admin = vm.addr(uint256(keccak256("Admin")));
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/invariants/interop/Setup.sol]: Isn't this the proxy admin?

We need to update the var name, but also to deploy the `ProxyAdmin` since it is a contract and not an EOA

File: packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { Setup, Constants, ConfigType } from "./Setup.sol";
+
+contract FuzzTest is Setup {
+    function test_place_holder() public {
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol]: We need to tests that the code length of the deployed contracts is greater zero and/or that the initialized vars are correct to make sure the setup is working

File: packages/contracts-bedrock/test/invariants/interop/Setup.sol (hunk 2 above)
Comment (hexshire) [reply packages/contracts-bedrock/test/invariants/interop/Setup.sol]: Does it change anything to our test if the Admin is a contract or a EOA?

Comment (hexshire) [reply packages/contracts-bedrock/test/invariants/interop/Setup.sol]: I mean, we are not testing the proxies and CHUGSPLASH and ERC1967 store the implementation in the same slot and both use delegatecall. Not sure why we would need to use ProxyAdmin contract since we are setting the implementation and admin with `vm.store`.

File: packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol (hunk 3 above)
Comment (hexshire) [reply packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol]: Added the code.length check.

File: packages/contracts-bedrock/medusa-debug.json [hunk 4]
```diff
@@ -0,0 +1,96 @@
+{
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/medusa-debug.json:1]: File to be removed afterwards, only created to debug faster in medusa w/o wasting too much time

File: packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import { Setup, Constants, ConfigType, GameType } from "./Setup.sol";
+import { Helpers } from "./utils/Helpers.sol";
+
+contract FuzzTest is Setup {
+    using Helpers for *;
+
+    bool initialized;
+
+    /// NOTE: Using this modifier because the initialization is not working when called inside the constructor on medusa
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol:20]: As this comment explains, `initialize()` call didn't revert but the state didn't persist when calling it on the constructor -  that's why we need to use this modifier

File: packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    modifier isInitialized() {
+        if (!initialized) {
+            _initializeEverything();
+            initialized = true;
+            _;
+        }
+    }
+
+    /// @notice Tests the contracts vars are set up correctly
+    function test_setup() public isInitialized {
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol]: This test will probably be set as an internal function assertion inside the constructor since it only checks the setup, and doesn't need to be fuzzed multiple times. Leaving here for now to check that Medusa works and it passes (it does)

Comment (simon-something) [reply packages/contracts-bedrock/test/invariants/interop/FuzzTest.t.sol]: We can even consider removing it at the end, it's rather useful to debug this (complex) setup - not sure how it's influencing fuzzing coverage tho (as it's kind of a coverage-dead end anyway)?

File: packages/contracts-bedrock/test/invariants/interop/Handler.t.sol [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import { Setup } from "./Setup.sol";
+import { Actors } from "./Actors.t.sol";
+
+contract Handler is Setup, Actors {
+    /// @notice Mint SUPER_TOKEN to an actor
+    /// @param _amount Amount to mint
+    function handler_mintSuperchainERC20(uint256 _amount) external withActor(msg.sender) {
+        _amount = clampLte(_amount, type(uint256).max - SUPER_TOKEN.totalSupply());
+        SUPER_TOKEN.mint(currentActor(), _amount);
+    }
```
Comment (simon-something) [inline packages/contracts-bedrock/test/invariants/interop/Handler.t.sol]: should we fuzz balances or perhaps mint for every actor in the setup?

File: packages/contracts-bedrock/medusa-debug.json (hunk 4 above)
Comment (hexshire) [reply packages/contracts-bedrock/medusa-debug.json:1]: We have to remove the command from the `justfile` too.

File: packages/contracts-bedrock/test/invariants/interop/Handler.t.sol (hunk 7 above)
Comment (0xDiscotech) [reply packages/contracts-bedrock/test/invariants/interop/Handler.t.sol]: I can move it on the setup, but should we remove this handler function?

**Comment by 0xDiscotech** [comment]
NOTE:
fails : `1` on try -> weird, since it works well on the catch when insufficient balance, meaning that it does call the contract as expected. But it fails on the assertions as if the call failed
uncovered: `3` on try

File: packages/contracts-bedrock/test/invariants/interop/Handler.t.sol (hunk 7 above)
Comment (0xDiscotech) [reply packages/contracts-bedrock/test/invariants/interop/Handler.t.sol]: Added on setup and removed from handler on https://github.com/defi-wonderland/optimism/pull/168/commits/8acfed8b8c30a2c2d2bed7634ad803ee024ff8bb

File: packages/contracts-bedrock/test/invariants/interop/helpers/Actors.t.sol [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        returns (bool)
+    {
+        try ISuperchainTokenBridge(superchainTokenBridge).relayERC20(_token, _from, _to, _amount) {
+            return true;
+        } catch {
+            return false;
+        }
+    }
+
+    function callBridgeSendERC20(
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/invariants/interop/helpers/Actors.t.sol:40]: Tried to use a function to do a low-level call with the input data to the superchain token bridge address, but some calls failed (weird). 
What happened is that it called the function and returned `success = true` but under the hood the function was never reached. A part of the log:
```
                 => [event] ActorsLog("call using actor: 0x1BaA33259E533141Ff86A4593d3b47Ff9A209f43")
                 => [call] StdCheats.toString(address)(0x4200000000000000000000000000000000000028) (addr=0x7109709ECfa91a80626fF3989D68f67F5b1DD12D, value=0, sender=0x1BaA33259E533141Ff86A4593d3b47Ff9A209f43)
                         => [return ("0x4200000000000000000000000000000000000028")]
                 => [event] ActorsLog("stoken bridge address: 0x4200000000000000000000000000000000000028")
                 => [call] <unresolved contract>.<unresolved method>(msg_data=c1a433d80000000000000000000000007ba37b1519173c980d34f1a4c508ebc495fb28fd00000000000000000000000001000000000000000000000000000000000000ff0000000000000000000000000000000091bf83ff9e537434a0fb488291a1864d0000147129c78276f6b18a880bab012a955597db9824526c73bb5350d3d87df1) (addr=0x4200000000000000000000000000000000000028, value=0, sender=0x1BaA33259E533141Ff86A4593d3b47Ff9A209f43)
                         => [return]
                 => [call] StdCheats.toString(bytes)() (addr=0x7109709ECfa91a80626fF3989D68f67F5b1DD12D, value=0, sender=0x1BaA33259E533141Ff86A4593d3b47Ff9A209f43)
                         => [return ("0x")]
                 => [event] ActorsLog("0x")
                 => [return (true, )]
```

I even copy pasted the msg data and tried it on a foundry test, and the call reached the function as expected.
Luckily, this behavior stopped when using high level calls (that's why I'm using them in this way), and I tried to mimic the return true and false behavior inside the try catch to handle the result on the properties tests.

File: packages/contracts-bedrock/test/invariants/interop/Setup.sol [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    address public superchainConfigAddress = vm.addr(uint256(keccak256("SuperchainConfig")));
+    address public systemConfigAddress = vm.addr(uint256(keccak256("SystemConfig")));
+    address public optimismPortalAddress = vm.addr(uint256(keccak256("OptimismPortal")));
+
+    // Internals
+    address internal _disputeGameFactory = vm.addr(uint256(keccak256("DisputeGameFactory")));
+    bytes internal _proxyCode;
+
+    constructor() {
+        vm.chainId(CHAIN_ID_ONE);
```
Comment (hexshire) [inline packages/contracts-bedrock/test/invariants/interop/Setup.sol:85]: Not necessary since the default value is 1, but is true that is more explicit like this.

**Review by hexshire — DISMISSED** [review]

**Review by hexshire — APPROVED** [review]

File: packages/contracts-bedrock/medusa-debug.json [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    "coverageFormats": ["html", "lcov"],
+    "targetContracts": ["FuzzTest"],
+    "predeployedContracts": {
+      "Deployer815": "0x4200000000000000000000000000000000000815",
+      "Deployer825": "0x4200000000000000000000000000000000000825"
+    },
+    "targetContractsBalances": [],
+    "constructorArgs": {},
+    "deployerAddress": "0x30000",
+    "senderAddresses": ["0x10000"],
```
Comment (simon-something) [inline packages/contracts-bedrock/medusa-debug.json:20]: so we don't use this to really fuzz, right? I mean, that's only one actor then

**Review by simon-something — APPROVED** [review]
🚀

File: packages/contracts-bedrock/medusa-debug.json (hunk 10 above)
Comment (0xDiscotech) [reply packages/contracts-bedrock/medusa-debug.json:20]: this is just for a debug script we run to check there is no an error on the tests easily, but it differs from the `medusa.json` one which is used for fuzzing more
